README.md
# BubbleChart

[![Build Status (master,develop)](https://travis-ci.org/jondavidjohn/bubblechart.png?branch=master,develop)](https://travis-ci.org/jondavidjohn/bubblechart)

BubbleChart is a JavaScript module for comparative visualization of two dimensional data in a bubble chart.

## Usage

To Create a new chart you simple create a new `BubbleChart` Object passing it
your configuration options, then to create the chart, just call `.paint()`

```js
var chart = new BubbleChart({
  canvasId: "bubblechart",
  metric: "Kb",
  usedArea: 0.35,
  contain: true,
  popoverOpts: {
    textFont: "Open Sans",
  },
  data : myCustomData()
});
chart.paint();
```

### Options

This is the exploded view of possbile options (with included defaults)

```js
{
  canvasId: "",      // id of target <canvas>
  usedArea: 0.5,     // how much of the canvas you wish to use for bubbles
  data: [],          // Array of data objects (converted into bubbles)
  metric: "",        // Label for your data metric (shown in popover)
  attribution: true, // controls "(Powered by BubbleChart)"
  fillColors: [],    // Array of colors to randomly choose from if data
                     // object lacks fillColor

  contain: false,       // keep bubbles within the canvas
  gutter: 0,            // containment gutter
  borderColor: "",      // Global Bubble border color
  borderSize: 0,        // Global Bubble border size
  textColor: "#fff"     // Bubble label color
  textType: "helvetica" // Font

  popoverOpts: {           // Control the look of the hover popover
    textFont: 'helvetica',
    textColor: '#fff',
    fillColor: '#333',
    labelSize: 22,         // Popover Heading size
    metricSize: 14,        // Popover Metric size
    opacity: 0.6,
  }
}
```

### Data Objects

Data objects are what create the bubbles.  Any of these objects will cascade
options set globally in the main BubbleChart Object.

Each data object's `data` property will determine the size of the bubble in
relation to the rest of the data object's `data` properties.  This is how you
get the visual representation of your data.

```js
{
  label: "Name of the thing",
  data: 10,
  metric: "this specific metric",
  borderColor: "#FFF",
  borderSize: 3,
  img_src: 'http://.../someImage.jpg'
}
```

#### Images

You can provide an image to be centered within the bubble by providing
a `img_src` property in your data objects.  **Keep in mind the following:**

  - Images are expected to be square, it will try to be smart and crop the
  image respecting width over height, but providing square images will take
  the guess work out of it.
  - Images will be used at full size if the width of the image is less than
  80% of the bubble diameter.
  - Bubbles will be sized down to fit no more than 80% of the bubble area
  - You can specify a `img_area` to adjust the display

Example of a data object with an image covering 60% of the bubble:

```js
{
  label: "Name of the thing",
  data: 10,
  img_src: 'http://.../image.jpg'
  img_area: 0.6
}
```

##### Retina

If your `img_src` value contains `@2x` it will be treated as a retina image and
be downsized (by 2) for display.

If you do not specify `@2x` images, and the client pixel density is greater
than 1, this module will reduce the image size to prefer image clarity over size.

#### Loading Spinner

There is a provided Spinner object you can use to show a loading animation if
loading your custom data is a time consuming process.

Example

```js
var chart = new BubbleChart({
  canvasId: "bubblechart",
  metric: "Kb",
  usedArea: 0.35,
  contain: true,
  popoverOpts: {
    textFont: "Open Sans",
  }
});

chart.spinner.start();
chart.data = myCustomData();
chart.reload()
chart.spinner.stop();
chart.paint();
```

## Requirements

  - IE9+ (required `<canvas>` support)
  - **No reliance on any external JavaScript Library**

## Development

see [CONTRIBUTING.md](https://github.com/jondavidjohn/bubblechart/blob/develop/CONTRIBUTING.md)



CHANGELOG
= 1.1.2
  * Remove inconsistent test (failing in travis / succeeding locally)

= 1.1.1
  * Configurable tooltip font sizes
  * Fix mouse position for retina displays - Issue #1
  * Explicit support for retina (@2x) images
  * Better Retina support and consistency in general

= 1.1.0
  * Move the requestAnimationFrame and general performance enhancements
  * Implement Images support in bubbles
  * Add Spinner control for longer data loading times

= 1.0.1
  * Fixed issue with accidental link following

= 1.0.0 =
  * Initial Release


test/PatchesTest.js
(function() {
	module('BubbleChart Patches Tests');

	test('randInt', function() {
		var test5 = Math.randInt(5),
			test3 = Math.randInt(3),
			test8 = Math.randInt(8);

		equal(true, test5 < 5);
		equal(true, test5 >= 0);
		equal(true, test3 < 3);
		equal(true, test3 >= 0);
		equal(true, test8 < 8);
		equal(true, test8 >= 0);
	});
})();


test/classes/BubbleChartTest.js
(function() {
	module('BubbleChart Tests');

	function generate(override_opts) {
		var opt,
			opts = {
				canvasId: 'test_canvas',
				data: [{
					label: "thing10",
					data: 10
				},{
					label: "thing5",
					data: 5
				},{
					label: "thing35",
					data:35
				},{
					label: "thing",
					data: 10
				}],
				metric: 'metric',
				fillColors: ['a', 'b', 'c', 'd']
			};

		for (opt in override_opts) { opts[opt] = override_opts[opt]; }
		return new BubbleChart(opts);
	}

	test('constructor', function() {
		var chart = generate();
		equal(chart.canvas.id, 'test_canvas');
		equal(chart.data.length, chart.bubbles.length);
		equal(chart.metric, 'metric');
		equal(chart.fillColors.length, 4);
		equal(chart.metricTotal, 60);
	});

	test('getPixelRatio', function() {
		var canvas = document.getElementById('test_canvas');

		canvas.context = canvas.getContext('2d');

		window.devicePixelRatio = 2;
		canvas.context.webkitBackingStorePixelRatio = 1;
		equal(BubbleChart.getPixelRatio(canvas.context), window.devicePixelRatio);

		window.devicePixelRatio = 2;
		canvas.context.webkitBackingStorePixelRatio = 2;
		equal(BubbleChart.getPixelRatio(canvas.context), 1);
	});


	test('attribution-before-default', function() {
		var chart = generate();
		notStrictEqual(chart.canvas.previousSibling.innerHTML.indexOf('https://github.com/jondavidjohn/bubblechart'), -1);
	});
	test('attribution-before', function() {
		var chart = generate({ attribution: 'before' });
		notStrictEqual(chart.canvas.previousSibling.innerHTML.indexOf('https://github.com/jondavidjohn/bubblechart'), -1);
	});
	test('attribution-after', function() {
		var chart = generate({ attribution: 'after' });
		notStrictEqual(chart.canvas.nextSibling.innerHTML.indexOf('https://github.com/jondavidjohn/bubblechart'), -1);
	});
	test('attribution-off', function() {
		var chart = generate({ attribution: false }),
			prev_sibling = chart.canvas.previousSibling,
			next_sibling = chart.canvas.nextSibling,
			assertions = 0;



		if ('innerHTML' in next_sibling) {
			assertions++;
			equal(next_sibling.innerHTML.indexOf('https://github.com/jondavidjohn/bubblechart'), -1);
		}

		if ('innerHTML' in prev_sibling) {
			assertions++;
			equal(prev_sibling.innerHTML.indexOf('https://github.com/jondavidjohn/bubblechart'), -1);
		}

		expect(assertions);
	});
})();


test/classes/BubbleChartTests.js
module("BubbleChart Tests");

test("truthy", function() {
  ok(true, "true is truthy");
  equal(1, true, "1 is truthy");
  notEqual(0, true, "0 is NOT truthy");
});


javascripts/congress.js
(function() {
	var chart = new BubbleChart({
			canvasId: "congress-canvas",
			metric: "% - Non-votes",
			usedArea: 0.4,
			attribution: false,
			contain: true,
			popoverOpts: {
				textFont: "Raleway",
			}
		}),
		rm = new RequestManager(),
		processVotes = null,
		nonvote_data = {},
		totalvote_data = {},
		chart_data = [];

	chart.spinner.start();

	function _url(endpoint, params) {
		var base_url = 'http://congress.api.sunlightfoundation.com/',
			apikey = congress_key,
			params_qs = [];

		params = params || {};
		params.apikey = apikey;

		for (var param in params) {
			if (params.hasOwnProperty(param)) {
				params_qs.push(param + '=' + params[param]);
			}
		}

		return base_url + endpoint + '?' + params_qs.join('&');
	}

	function formatChartData(bioguide_id, stat) {
		var leg;
		// Make sure we're dealing with someone still in office
		if (legislators.hasOwnProperty(bioguide_id) && legislators[bioguide_id].in_office === "1") {
			leg = legislators[bioguide_id];
			return {
				label: leg.title + '. ' + leg.firstname + ' ' + leg.lastname + ' (' + leg.party + '-' + leg.state + ')',
				data: stat,
				fillColor: getPartyColor(leg.party),
				href: leg.congresspedia_url,
				img_src: '/bubblechart/images/leg/' + bioguide_id + '.jpg',
			};
		}
	}

	processVotes = function(res) {
		var i, id, len, stat;

		// For each vote
		for (i = 0, len = res.results.length; i < len; i++) {
			// Loop through all voters
			for (id in res.results[i].voter_ids) {
				if (res.results[i].voter_ids.hasOwnProperty(id)) {
					if (["Not Voting", "Present"].indexOf(res.results[i].voter_ids[id]) !== -1) {
						if (nonvote_data.hasOwnProperty(id)) {
							nonvote_data[id]++;
						}
						else {
							nonvote_data[id] = 1;
						}
					}

					if (totalvote_data.hasOwnProperty(id)) {
						totalvote_data[id]++;
					}
					else {
						totalvote_data[id] = 1;
					}
				}
			}
		}

		if (!rm.outstandingRequests()) {
			for (id in nonvote_data) {
				if (nonvote_data.hasOwnProperty(id)) {
					stat = Math.round((nonvote_data[id] / totalvote_data[id]) * 100);
					if (stat > 15) {
						chart_data.push(formatChartData(id, stat));
					}
				}
			}

			chart_data = chart_data.filter(function(a) { return !!a; });

			chart.data = chart_data;
			chart.reload();
			chart.spinner.stop();
			chart.paint();
		}
	};

	// Initial request to get pagination info
	window.onload = function() {
		rm.execJSONP(
			_url('votes', {
				year: 2013,
				fields: 'voter_ids,roll_id',
				per_page: 50,
				page: 1
			}),
			function(res) {
				var pages = Math.floor(res.count / 50) + 1;

				for (var i = 2; i <= pages; i++) {
					rm.execJSONP(
						_url('votes', {
							year: 2013,
							fields: 'voter_ids,roll_id',
							per_page: 50,
							page: i
						}),
						processVotes
					);
				}

				processVotes(res); // Process page 1
			}
		);
	};

	function getPartyColor(party) {
		switch(party) {
			case 'R':
				return '#861316';

			case 'D':
				return '#033A76';

			default:
				return '#bbb';
		}
	}
}).call();


javascripts/bcexamples.js
(function() {
	var request_count = 0;
	var access_token = 'dcfbecfd0d0d2a6d28b82d96dca49da93aec6e3e';

	window.BCExamples = {
		execJSONP: function(url, callback) {
			var script = document.createElement('script'),
				randCbName = 'bc_' + Math.floor(Math.random() * 65535) + 1,
				sepChar = (url.indexOf('?') === -1) ? '?' : '&';

			request_count++;
			if (access_token.length) {
				url = url + sepChar + 'access_token=' + access_token;
				sepChar = (url.indexOf('?') === -1) ? '?' : '&';
			}
			script.async = true;
			window[randCbName] = function(data) {
				request_count--;
				var scr = document.getElementById(randCbName);
				scr.parentNode.removeChild(scr);
				callback(data);
				window[randCbName] = null;
				delete window[randCbName];
			};

			script.src = url+sepChar+'callback='+randCbName;
			script.id = randCbName;
			document.getElementsByTagName('head')[0].appendChild(script);
		},
		outstandingRequests: function() {
			return request_count > 0;
		},
		getLangColor: function(lang) {
			var colors = {
				"Arduino": "#bd79d1",
				"Java": "#b07219",
				"VHDL": "#543978",
				"Scala": "#7dd3b0",
				"Emacs Lisp": "#c065db",
				"Delphi": "#b0ce4e",
				"Ada": "#02f88c",
				"VimL": "#199c4b",
				"Perl": "#0298c3",
				"Lua": "#fa1fa1",
				"Rebol": "#358a5b",
				"Verilog": "#848bf3",
				"Factor": "#636746",
				"Ioke": "#078193",
				"R": "#198ce7",
				"Erlang": "#949e0e",
				"Nu": "#c9df40",
				"AutoHotkey": "#6594b9",
				"Clojure": "#db5855",
				"Shell": "#5861ce",
				"Assembly": "#a67219",
				"Parrot": "#f3ca0a",
				"C#": "#5a25a2",
				"Turing": "#45f715",
				"AppleScript": "#3581ba",
				"Eiffel": "#946d57",
				"Common Lisp": "#3fb68b",
				"Dart": "#cccccc",
				"SuperCollider": "#46390b",
				"CoffeeScript": "#244776",
				"XQuery": "#2700e2",
				"Haskell": "#29b544",
				"Racket": "#ae17ff",
				"Elixir": "#6e4a7e",
				"HaXe": "#346d51",
				"Ruby": "#701516",
				"Self": "#0579aa",
				"Fantom": "#dbded5",
				"Groovy": "#e69f56",
				"C": "#555",
				"JavaScript": "#f15501",
				"D": "#fcd46d",
				"ooc": "#b0b77e",
				"C++": "#f34b7d",
				"Dylan": "#3ebc27",
				"Nimrod": "#37775b",
				"Standard ML": "#dc566d",
				"Objective-C": "#f15501",
				"Nemerle": "#0d3c6e",
				"Mirah": "#c7a938",
				"Boo": "#d4bec1",
				"Objective-J": "#ff0c5a",
				"Rust": "#dea584",
				"Prolog": "#74283c",
				"Ecl": "#8a1267",
				"Gosu": "#82937f",
				"FORTRAN": "#4d41b1",
				"ColdFusion": "#ed2cd6",
				"OCaml": "#3be133",
				"Fancy": "#7b9db4",
				"Pure Data": "#f15501",
				"Python": "#3581ba",
				"Tcl": "#e4cc98",
				"Arc": "#ca2afe",
				"Puppet": "#cc5555",
				"Io": "#a9188d",
				"Max": "#ce279c",
				"Go": "#8d04eb",
				"ASP": "#6a40fd",
				"Visual Basic": "#945db7",
				"PHP": "#6e03c1",
				"Scheme": "#1e4aec",
				"Vala": "#3581ba",
				"Smalltalk": "#596706",
				"Matlab": "#bb92ac",
				"C#": "#bb92af"
			};
			return colors[lang];
		}
	};
})();


javascripts/gh-lang.js
(function() {
	var chart = new BubbleChart({
			canvasId: "gh-lang-canvas",
			metric: "Kb",
			usedArea: 0.35,
			contain: true,
			attribution: false,
			popoverOpts: {
				textFont: "Raleway",
			}
		}),
		lang_data = {},
		rm = new RequestManager(),
		timer;

	function _url(endpoint, params) {
		var base_url = 'https://api.github.com/',
			access_token = gh_key,
			params_qs = [];

		params = params || {};
		params.access_token = access_token;

		for (var param in params) {
			params_qs.push(param + '=' + params[param]);
		}

		if (endpoint.substring(0,4) == 'http') {
			return endpoint + '?' + params_qs.join('&');
		}
		else {
			return base_url + endpoint + '?' + params_qs.join('&');
		}
	}

	document.getElementById('gh-lang-username-btn').onclick = function(e) {
		var input = document.getElementById('gh-lang-username'),
			button = e.target || e.srcElement,
			username = input.value;

		input.disabled = true;
		button.disabled = true;
		button.className = button.className += ' disabled';
		chart.spinner.start();

		rm.execJSONP(
			_url('users/'+username+'/repos'),
			function(response) {
				if (response.meta.status !== 200) {
					alert(response.data.message);
					return;
				}

				for (var i = 0, len = response.data.length; i < len; i++) {
					if (response.data[i].fork) continue;

					(function(repo) {
						rm.execJSONP(
							_url(repo.languages_url),
							function(lang_response) {
								for (var lang in lang_response.data) {
									(function(lang, bytes) {
										if (!lang_data.hasOwnProperty(lang)) {
											lang_data[lang] = 0;
										}
										lang_data[lang] += bytes;
									})(lang, lang_response.data[lang]);
								}
							}
						);
					})(response.data[i]);
				}
			}
		);

		setTimeout(function(){
			timer = setInterval(function() {
				if (rm.outstandingRequests()) return;

				clearInterval(timer);

				chart.data = [];

				for (var lang in lang_data) {
					(function(lang, bytes) {
						chart.data.push({
							label: lang,
							data: Math.round(bytes / 1024), // kb
							fillColor: langColor(lang)
						});
					})(lang, lang_data[lang]);
				}

				if (chart.data.length) {
					chart.spinner.stop();
					chart.reload();
					chart.paint();
				}

				lang_data = {};
				input.disabled = false;
				button.disabled = false;
				button.className = button.className.replace('disabled', '');
			}, 1000);
		}, 500);
	};

	function langColor(lang) {
		var colors = {
			"Arduino": "#bd79d1",
			"Java": "#b07219",
			"VHDL": "#543978",
			"Scala": "#7dd3b0",
			"Emacs Lisp": "#c065db",
			"Delphi": "#b0ce4e",
			"Ada": "#02f88c",
			"VimL": "#199c4b",
			"Perl": "#0298c3",
			"Lua": "#fa1fa1",
			"Rebol": "#358a5b",
			"Verilog": "#848bf3",
			"Factor": "#636746",
			"Ioke": "#078193",
			"R": "#198ce7",
			"Erlang": "#949e0e",
			"Nu": "#c9df40",
			"AutoHotkey": "#6594b9",
			"Clojure": "#db5855",
			"Shell": "#5861ce",
			"Assembly": "#a67219",
			"Parrot": "#f3ca0a",
			"Turing": "#45f715",
			"AppleScript": "#3581ba",
			"Eiffel": "#946d57",
			"Common Lisp": "#3fb68b",
			"Dart": "#cccccc",
			"SuperCollider": "#46390b",
			"CoffeeScript": "#244776",
			"XQuery": "#2700e2",
			"Haskell": "#29b544",
			"Racket": "#ae17ff",
			"Elixir": "#6e4a7e",
			"HaXe": "#346d51",
			"Ruby": "#701516",
			"Self": "#0579aa",
			"Fantom": "#dbded5",
			"Groovy": "#e69f56",
			"C": "#555",
			"JavaScript": "#f15501",
			"D": "#fcd46d",
			"ooc": "#b0b77e",
			"C++": "#f34b7d",
			"Dylan": "#3ebc27",
			"Nimrod": "#37775b",
			"Standard ML": "#dc566d",
			"Objective-C": "#f15501",
			"Nemerle": "#0d3c6e",
			"Mirah": "#c7a938",
			"Boo": "#d4bec1",
			"Objective-J": "#ff0c5a",
			"Rust": "#dea584",
			"Prolog": "#74283c",
			"Ecl": "#8a1267",
			"Gosu": "#82937f",
			"FORTRAN": "#4d41b1",
			"ColdFusion": "#ed2cd6",
			"OCaml": "#3be133",
			"Fancy": "#7b9db4",
			"Pure Data": "#f15501",
			"Python": "#3581ba",
			"Tcl": "#e4cc98",
			"Arc": "#ca2afe",
			"Puppet": "#cc5555",
			"Io": "#a9188d",
			"Max": "#ce279c",
			"Go": "#8d04eb",
			"ASP": "#6a40fd",
			"Visual Basic": "#945db7",
			"PHP": "#6e03c1",
			"Scheme": "#1e4aec",
			"Vala": "#3581ba",
			"Smalltalk": "#596706",
			"Matlab": "#bb92ac",
			"C#": "#bb92af"
		};
		return colors[lang];
	}
}).call();
